src/librustc/ty/query/mod.rs

```diff
@@ -97,228 +98,228 @@ pub use self::on_disk_cache::OnDiskCache;
     // might want to use `reveal_all()` method to change modes.
-    [] fn param_env: ParamEnv(DefId) -> ty::ParamEnv<'tcx>,
+    [] category<TypeChecking> fn param_env: ParamEnv(DefId) -> ty::ParamEnv<'tcx>,
 
     // Trait selection queries. These are best used by invoking `ty.moves_by_default()`,
     // `ty.is_copy()`, etc, since that will prune the environment where possible.
-    [] fn is_copy_raw: is_copy_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
-    [] fn is_sized_raw: is_sized_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
-    [] fn is_freeze_raw: is_freeze_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
-    [] fn needs_drop_raw: needs_drop_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
-    [] fn layout_raw: layout_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>)
+    [] category<TypeChecking> fn is_copy_raw: is_copy_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
+    [] category<TypeChecking> fn is_sized_raw: is_sized_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
+    [] category<TypeChecking> fn is_freeze_raw: is_freeze_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
+    [] category<TypeChecking> fn needs_drop_raw: needs_drop_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool,
+    [] category<TypeChecking> fn layout_raw: layout_dep_node(ty::ParamEnvAnd<'tcx, Ty<'tcx>>)
                                   -> Result<&'tcx ty::layout::LayoutDetails,
                                             ty::layout::LayoutError<'tcx>>,
 
-    [] fn dylib_dependency_formats: DylibDepFormats(CrateNum)
+    [] category<Other> fn dylib_dependency_formats: DylibDepFormats(CrateNum)
                                     -> Lrc<Vec<(CrateNum, LinkagePreference)>>,
 
-    [fatal_cycle] fn is_panic_runtime: IsPanicRuntime(CrateNum) -> bool,
-    [fatal_cycle] fn is_compiler_builtins: IsCompilerBuiltins(CrateNum) -> bool,
-    [fatal_cycle] fn has_global_allocator: HasGlobalAllocator(CrateNum) -> bool,
-    [fatal_cycle] fn is_sanitizer_runtime: IsSanitizerRuntime(CrateNum) -> bool,
-    [fatal_cycle] fn is_profiler_runtime: IsProfilerRuntime(CrateNum) -> bool,
-    [fatal_cycle] fn panic_strategy: GetPanicStrategy(CrateNum) -> PanicStrategy,
-    [fatal_cycle] fn is_no_builtins: IsNoBuiltins(CrateNum) -> bool,
+    [fatal_cycle] category<Codegen> fn is_panic_runtime: IsPanicRuntime(CrateNum) -> bool,
+    [fatal_cycle] category<Codegen> fn is_compiler_builtins: IsCompilerBuiltins(CrateNum) -> bool,
+    [fatal_cycle] category<Codegen> fn has_global_allocator: HasGlobalAllocator(CrateNum) -> bool,
+    [fatal_cycle] category<Codegen> fn is_sanitizer_runtime: IsSanitizerRuntime(CrateNum) -> bool,
+    [fatal_cycle] category<Codegen> fn is_profiler_runtime: IsProfilerRuntime(CrateNum) -> bool,
+    [fatal_cycle] category<Codegen> fn panic_strategy: GetPanicStrategy(CrateNum) -> PanicStrategy,
+    [fatal_cycle] category<Codegen> fn is_no_builtins: IsNoBuiltins(CrateNum) -> bool,
```

Comment:
Answer: yes, we are still using `[]` for *something*. Maybe we should change that to `#[fatal_cycle]` as well.

Reply:
I kind of like that idea, but it seems a little surprising to me that in this context `#[whatever]` doesn't mean what it normally would. For example, `#[derive(Clone)]` wouldn't work.

Reply:
I think this *is* what `#[...]` normally means. That is, `#[..]` is some kind of meta-data interpreted by the context. In this context, this is a macro, and macros often re-interpret `#[foo]` annotations. (Consider for example `#[derive]` which allows for annotations on fields etc that aren't normally used.) That doesn't mean that all attributes would be appropriate of course (e.g. `#[derive(Clone)]` doesn't make sense on a function anyhow)

Reply:
Fair enough! I don't think I've seen a macro re-interpret `#[]` annotations before so it seemed strange to me. I can certainly do this though.

Reply:
Interestingly enough, it looks like `define_queries!` [already](https://github.com/rust-lang/rust/blob/01cc982e936120acb0424e41de14e42ba2d88c6f/src/librustc/ty/query/plumbing.rs#L635) [handles](https://github.com/rust-lang/rust/blob/01cc982e936120acb0424e41de14e42ba2d88c6f/src/librustc/ty/query/plumbing.rs#L651) [attributes](https://github.com/rust-lang/rust/blob/01cc982e936120acb0424e41de14e42ba2d88c6f/src/librustc/ty/query/plumbing.rs#L686). This doesn't appear to be used currently or in the recent history of this file. @nikomatsakis would you like me to continue with this idea and remove support for regular `#[thing]` attributes from the macro? What do you think of @eddyb's [idea](https://github.com/rust-lang/rust/pull/51657#issuecomment-399310003)? I quite like it myself.

Reply:
I'm fine with @eddyb's idea too